### PR TITLE
Moving Microsoft.Extensions.Hosting* from Extensions to Runtime

### PR DESCRIPTION
**Pull request by Anipik**
Related to https://github.com/dotnet/runtime/issues/46091

- Move the src projects
- Generate the ref source
- Add the ref project
- Formatting changes
- Add pkgprojs
- Add slns 

Left
- Add test projs

### CODE REVIEW

File: src/libraries/pkg/baseline/packageIndex.json [hunk 1]
```diff
@@ -772,6 +772,56 @@
...
+        "3.1.2",
+        "3.1.3",
+        "3.1.4",
+        "3.1.5",
+        "3.1.6",
+        "3.1.7",
+        "3.1.8",
+        "3.1.9",
+        "3.1.10",
+        "5.0.0"
```
Comment (eerhardt) [inline src/libraries/pkg/baseline/packageIndex.json:792]: It looks like there are `5.0.1` versions of both of these packages on nuget.org.

[https://www.nuget.org/packages/Microsoft.Extensions.Hosting.WindowsServices](https://www.nuget.org/packages/Microsoft.Extensions.Hosting.WindowsServices)

File: src/libraries/Microsoft.Extensions.Hosting.Systemd/ref/Microsoft.Extensions.Hosting.Systemd.csproj [hunk 2]
```diff
@@ -0,0 +1,12 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFrameworks>netstandard2.1</TargetFrameworks>
+    <EnableDefaultItems>true</EnableDefaultItems>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="$(LibrariesProjectRoot)Microsoft.Extensions.Hosting\src\Microsoft.Extensions.Hosting.csproj" />
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Hosting.Systemd/ref/Microsoft.Extensions.Hosting.Systemd.csproj:9]: This is the `ref` project. Shouldn't it be referencing the `ref` project of Hosting?

File: src/libraries/pkg/baseline/packageIndex.json [hunk 3]
```diff
@@ -772,6 +772,56 @@
         "6.0.0.0": "6.0.0"
       }
     },
+    "Microsoft.Extensions.Hosting.Systemd": {
```
Comment (eerhardt) [inline src/libraries/pkg/baseline/packageIndex.json:775]: Where do things like

```xml
<PackageTags>hosting</PackageTags>
```

go in our build infrastructure? Both packages included that tag before.

File: src/libraries/Microsoft.Extensions.Hosting.WindowsServices/src/Internal/Win32.cs [hunk 4]
```diff
@@ -0,0 +1,92 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+
+namespace Microsoft.Extensions.Hosting.WindowsServices.Internal
+{
+    internal static class Win32
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Hosting.WindowsServices/src/Internal/Win32.cs:10]: For this PR it is probably best to bring it over as-is. But we should follow up with moving these into the typical "Interop" pattern we have established in dotnet/runtime/src/libraries.

Comment (Anipik) [reply src/libraries/Microsoft.Extensions.Hosting.WindowsServices/src/Internal/Win32.cs:10]: > For this PR it is probably best to bring it over as-is. But we should follow up with moving these into the typical "Interop" pattern we have established in dotnet/runtime/src/libraries

yes thats the plan

File: src/libraries/Microsoft.Extensions.Hosting.WindowsServices/src/Microsoft.Extensions.Hosting.WindowsServices.csproj [hunk 5]
```diff
@@ -0,0 +1,14 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFrameworks>netstandard2.0;netstandard2.1</TargetFrameworks>
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Hosting.WindowsServices/src/Microsoft.Extensions.Hosting.WindowsServices.csproj:4]: I don't see a reason why this needs to cross-compile for both `2.0` and `2.1`. There's no `#if`s or file inclusion/exclusions. Can we drop the `netstandard2.1` TFM?

**Review by eerhardt — APPROVED** [review]
The general structure looks good to me. I just have a question on why the Windows Service package needs to target ns2.1.

I'm no expert on the pkgproj stuff, so it would be good to get another set of eyes on it from someone who understands it.

File: src/libraries/Microsoft.Extensions.Hosting.Systemd/Directory.Build.Props [hunk 6]
```diff
@@ -0,0 +1,7 @@
+<Project>
```
Comment (ViktorHofer) [inline src/libraries/Microsoft.Extensions.Hosting.Systemd/Directory.Build.Props:1]: The file name should be `Directory.Build.props` (lower-case props).

**Review by ViktorHofer — APPROVED** [review]

File: src/libraries/Microsoft.Extensions.Hosting.Systemd/Directory.Build.Props (hunk 6 above)
Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.Hosting.Systemd/Directory.Build.Props:1]: nice catch!

**Comment by ericstj** [comment]
>  I just have a question on why the Windows Service package needs to target ns2.1.

> I'm no expert on the pkgproj stuff, so it would be good to get another set of eyes on it from someone who understands it.

No package expertise needed.  It's just a copy of https://github.com/dotnet/extensions/blob/1e9779404285d61b7d0ff2bf95f985ec137f53a9/src/Hosting/WindowsServices/src/Microsoft.Extensions.Hosting.WindowsServices.csproj#L5

Which you can see does this because of https://github.com/dotnet/extensions/commit/1264b9e2567027d2243f77f48cc08fdd576e3005

**Comment by eerhardt** [comment]
> Which you can see does this because of dotnet/extensions@1264b9e

On that commit, there is more API for netstandard2.1, and they also changed the references.

I still don't understand why we need to build for both `netstandard2.0` and `netstandard2.1`

**Comment by ericstj** [comment]
Presumably to not have a reference to BCL.AsyncInterfaces?  Regardless we shouldn't be changing this code in a port PR as we'll lose the history and reason behind such changes.  Ports should be bare minimum copy + get it building.

**Comment by eerhardt** [comment]
> Regardless we shouldn't be changing this code in a port PR

Let's open an issue to follow up. Just like a follow up issue is needed for the the interop code.

**Comment by Anipik** [comment]
cc @JunTaoLuo @Tratcher

**Comment by JunTaoLuo** [comment]
What specifically did you want us to take a look inside pkgproj? As far as I know, it's something that's used in runtime, but we don't use pkgproj in aspnetcore. But I took a quick look anyway and didn't spot anything that looks out of place.

**Comment by Anipik** [comment]
> What specifically did you want us to take a look inside pkgproj? As far as I know, it's something that's used in runtime, but we don't use pkgproj in aspnetcore. But I took a quick look anyway and didn't spot anything that looks out of place.

okay awesome thanks.

File: src/libraries/Microsoft.Extensions.Hosting.WindowsServices/ref/Microsoft.Extensions.Hosting.WindowsServices.csproj [hunk 7]
```diff
@@ -1,10 +1,14 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-    <TargetFrameworks>netstandard2.0;netstandard2.1</TargetFrameworks>
+    <TargetFrameworks>net461;netstandard2.0;netstandard2.1</TargetFrameworks>
     <EnableDefaultItems>true</EnableDefaultItems>
   </PropertyGroup>
 
+  <ItemGroup Condition="$(TargetFramework.StartsWith('net4'))">
```
Comment (ViktorHofer) [inline src/libraries/Microsoft.Extensions.Hosting.WindowsServices/ref/Microsoft.Extensions.Hosting.WindowsServices.csproj:8]: StartsWith should only be used if the tfm includes a suffix.

```suggestion
  <ItemGroup Condition="'$(TargetFramework)' == 'net461'">
```

File: src/libraries/Microsoft.Extensions.Hosting.WindowsServices/src/Microsoft.Extensions.Hosting.WindowsServices.csproj [hunk 8]
```diff
@@ -1,14 +1,21 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-    <TargetFrameworks>netstandard2.0;netstandard2.1</TargetFrameworks>
+    <TargetFrameworks>net461;netstandard2.0;netstandard2.1</TargetFrameworks>
     <EnableDefaultItems>true</EnableDefaultItems>
   </PropertyGroup>
 
+  <ItemGroup Condition="$(TargetFramework.StartsWith('net4'))">
```
Comment (ViktorHofer) [inline src/libraries/Microsoft.Extensions.Hosting.WindowsServices/src/Microsoft.Extensions.Hosting.WindowsServices.csproj:8]:
```suggestion
  <ItemGroup Condition="'$(TargetFramework)' == 'net461'">
```

File: src/libraries/Microsoft.Extensions.Hosting.WindowsServices/src/Microsoft.Extensions.Hosting.WindowsServices.csproj [hunk 9]
```diff
@@ -1,14 +1,21 @@
...
+    <Reference Include="System.ServiceProcess" />
+  </ItemGroup>
+
+  <ItemGroup Condition="!$(TargetFramework.StartsWith('net4'))">
```
Comment (ViktorHofer) [inline src/libraries/Microsoft.Extensions.Hosting.WindowsServices/src/Microsoft.Extensions.Hosting.WindowsServices.csproj:12]:
```suggestion
  <ItemGroup Condition="'$(TargetFramework)' != 'net461'">
```

**Pull request merged**